### PR TITLE
Refactor splash and login code

### PR DESCRIPTION
**Pull request by Maxr1998**

### CODE REVIEW

**Comment by Maxr1998** [comment]
@dkanada rebased this onto latest master.

File: app/src/main/java/com/dkanada/gramophone/ui/activities/SplashActivity.java [hunk 1]
```diff
@@ -32,40 +33,31 @@ protected void onCreate(Bundle savedInstanceState) {
     @Override
     protected void onResume() {
         super.onResume();
-        if (detectBatteryOptimization()) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && detectBatteryOptimization()) {
```
Comment (dkanada) [inline app/src/main/java/com/dkanada/gramophone/ui/activities/SplashActivity.java:36]: Is this check required even with the decorators below? I've never used them before, but they seem to just shift the problem around. Seems good for documentation though so I think it's fine.

File: app/src/main/java/com/dkanada/gramophone/ui/activities/LoginActivity.java [hunk 2]
```diff
@@ -106,20 +107,21 @@ public void onError(Exception exception) {
...
+                    PreferenceUtil.getInstance(LoginActivity.this).setUser(user);
+                    PreferenceUtil.getInstance(LoginActivity.this).setToken(token);
+
+                    Intent intent = new Intent(LoginActivity.this, MainActivity.class);
+                    intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
+                    startActivity(intent);
+                    finish();
                 } else {
-                    Toast.makeText(context, context.getResources().getString(R.string.error_version), Toast.LENGTH_SHORT).show();
+                    Toast.makeText(LoginActivity.this, getResources().getString(R.string.error_version), Toast.LENGTH_SHORT).show();
```
Comment (dkanada) [inline app/src/main/java/com/dkanada/gramophone/ui/activities/LoginActivity.java:124]: This is interesting, what's the benefit of `LoginActivity.class` over the reference?

File: app/src/main/java/com/dkanada/gramophone/ui/activities/SplashActivity.java (hunk 1 above)
Comment (Maxr1998) [reply app/src/main/java/com/dkanada/gramophone/ui/activities/SplashActivity.java:36]: The decorators basically say that you may not use the method below a certain sdk. So, the checks are still necessary. I just moved them to a more central space pretty much.

File: app/src/main/java/com/dkanada/gramophone/ui/activities/LoginActivity.java (hunk 2 above)
Comment (Maxr1998) [reply app/src/main/java/com/dkanada/gramophone/ui/activities/LoginActivity.java:124]: You save an extra reference. The `this` itself would be different at that place, but with the classname prefix you can specify which `this` you mean. In Kotlin, there's a similar construct where you write `this@LoginActivity`. The key advantage is better readability/clarity and less code.

**Pull request merged**